File: node/get-access-token.js
const ClientOAuth2 = require("client-oauth2"),
      util = require("util"),
      readline = require("readline"),
      fs = require("fs"),
      Path = require("path"),
      url = require("url");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (q) => new Promise((resolve) => rl.question(q,resolve));
const writeFile = util.promisify(fs.writeFile);

const state = Date.now();

const OAUTH_URL = 'https://auth.aweber.com/oauth2',
      TOKEN_URL = 'https://auth.aweber.com/oauth2/token',
      scopes = [
          'account.read',
          'list.read',
          'list.write',
          'subscriber.read',
          'subscriber.write',
          'email.read',
          'email.write',
          'subscriber.read-extended'
      ];



(async () => {

    let createRefresh = await question("Do you wish to create(c) tokens or refresh(r) tokens?:");
    let user;
    let credentials, clientId, clientSecret;

    if(createRefresh.toUpperCase() === "C") {

        clientId = await question("clientId:");
        clientSecret = await question("clientSecret:");

        const aweberAuth = new ClientOAuth2({
            clientId, clientSecret,
            accessTokenUri:TOKEN_URL,
            authorizationUri:`${OAUTH_URL}/authorize`,
            redirectUri:"http://localhost:3000/callback",
            scopes
        });
        const authorizationUrl = aweberAuth.code.getUri({state});
        console.log("Go to this url",authorizationUrl);

        const authorizationResponse = await question("Login and paste the returned URL here:");

        console.log("authorization response",authorizationResponse);
        user = await aweberAuth.code.getToken(authorizationResponse);

    } else if(createRefresh.toUpperCase() === "R") {

        credentials = require("./credentials.json");
        clientId = credentials.clientId;
        clientSecret = credentials.clientSecret;


        if(!credentials.accessToken ||
           !credentials.refreshToken ||
           !credentials.clientId ||
           !credentials.clientSecret) {
            throw new Error("credentials.json is missing fields");
        }

        const aweberAuth = new ClientOAuth2({
            clientId:credentials.clientId,
            clientSecret:credentials.clientSecret,
            accessTokenUri:TOKEN_URL,
            authorizationUri:`${OAUTH_URL}/authorize`,
            redirectUri:"http://localhost:3000/callback",
            scopes
        });
        user = await aweberAuth.createToken(
            credentials.accessToken,
            credentials.refreshToken,
            "bearer"
        ).refresh();
    }
    else {
        throw new Error("Invalid entry. You must enter 'c' or 'r'");
    }

    await writeFile(Path.join(__dirname,"credentials.json"),JSON.stringify({
        clientSecret:clientSecret,
        clientId:clientId,
        accessToken:user.accessToken,
        refreshToken:user.refreshToken
    }))
    process.exit();
})();

File: ruby/get_access_token.rb
require 'oauth2'
require "uri"
require "tty-prompt"

prompt = TTY::Prompt.new
createRefresh = prompt.ask("Do you wish to create(c) tokens or refresh(r) tokens?:")

SITE_URL = "https://auth.aweber.com"
AUTH_URL = '/oauth2/authorize'
TOKEN_URL = '/oauth2/token'
REDIRECT_URI = 'http://localhost:3000/callback'

scopes = [
    'account.read',
    'list.read',
    'list.write',
    'subscriber.read',
    'subscriber.write',
    'email.read',
    'email.write',
    'subscriber.read-extended'
]
SCOPE = scopes.join(" ")
puts SCOPE
class Encoder

  def self.encode(hash)
    URI.encode_www_form(hash)
  end

  def self.decode(string)
    URI.decode_www_form(hash)
  end
end
if createRefresh.upcase == "C"
    credentials = {}
    credentials[:client_id]  = prompt.ask("client id:")
    credentials[:client_secret] = prompt.ask("client secret:")

    client = OAuth2::Client.new(
        credentials[:client_id], 
        credentials[:client_secret], 
        :site => SITE_URL,
        :authorize_url => AUTH_URL,
        :token_url => TOKEN_URL,
        :connection_opts => {
          :request => {
            params_encoder: Encoder
          }
        },
        :scope => SCOPE
    )

    auth_url = client.auth_code.authorize_url(
      :redirect_uri => REDIRECT_URI,
      :scope => SCOPE
    )
    puts auth_url
    response_url = prompt.ask("visit the url above and paste the result here after you login:")

    code = CGI.parse(URI.parse(response_url).query)["code"].first

    puts "auth code is:"
    puts code

    token = client.auth_code.get_token(
        code,
        :redirect_uri => REDIRECT_URI
    )
    credentials[:access_token] = token.token
    credentials[:refresh_token] = token.refresh_token
else
    credentials = JSON.parse(File.read("./credentials.json"))

    client = OAuth2::Client.new(
        credentials["client_id"],
        credentials["client_secret"],
        :site => SITE_URL,
        :authorize_url => AUTH_URL,
        :token_url => TOKEN_URL,
        :scope => SCOPE
    )
    puts credentials["refresh_token"]
    token = OAuth2::AccessToken.new(
        client,
        credentials["access_token"],
        {
            refresh_token: credentials["refresh_token"],
            :scope => SCOPE
        }
    )
    new_token = token.refresh!(:scopes => SCOPE)

    credentials[:access_token] = new_token.token
    credentials[:refresh_token] = new_token.refresh_token

end

File.open("./credentials.json","w") do |f|
    f.write(credentials.to_json)
end

